// testcafe/lib/browser/provider/built-in/dedicated/chrome/cdp-client/shared-adapter-initializer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const execution_context_1 = __importDefault(require("./execution-context"));
const clients_manager_1 = require("./clients-manager");
const adapter_1 = require("../../../../../../shared/adapter");
adapter_1.initializeAdapter({
    PromiseCtor: Promise,
    nativeMethods: {
        setTimeout,
        clearTimeout,
        arrayIndexOf: Array.prototype.indexOf,
        arraySplice: Array.prototype.splice,
        arraySlice: Array.prototype.slice,
        arrayFilter: Array.prototype.filter,
        objectAssign: Object.assign,
        objectKeys: Object.keys,
        dateNow: Date.now,
    },
    isDomElement: () => false,
    scroll: async (el, opts) => {
        let currCxt = execution_context_1.default.current;
        let result = null;
        let margin = void 0;
        do {
            const { exceptionDetails, result: resultObj } = await clients_manager_1.getClient().Runtime.callFunctionOn({
                returnByValue: true,
                awaitPromise: true,
                executionContextId: execution_context_1.default.getCurrentContextId(),
                arguments: [{ objectId: el.objectId }, { value: opts }, { value: margin }],
                functionDeclaration: `function (el, opts) {
                    return window["%proxyless%"].scroll(el, opts);
                }`,
            });
            if (exceptionDetails)
                throw exceptionDetails;
            const scrollResult = resultObj.value;
            if (currCxt && currCxt !== currCxt.parent) {
                // TODO:
                //el           = findIframeByWindow(currCxt);
                currCxt = currCxt.parent;
                result = result !== null && result !== void 0 ? result : scrollResult.scrollWasPerformed;
                margin = scrollResult.maxScrollMargin;
                opts.offsetX = scrollResult.offsetX;
                opts.offsetY = scrollResult.offsetY;
            }
        } while (currCxt && currCxt !== currCxt.parent);
        return result;
    },
});
